src/main/java/org/wildfly/security/keystore/KeyStoreUtil.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.keystore;
+
+import java.io.DataInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.FileChannel;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.cert.CertificateException;
+import java.util.function.Supplier;
+
+import static org.wildfly.security._private.ElytronMessages.log;
+import static org.wildfly.security.util.ProviderUtil.findProvider;
+
+/**
+ * Utility functions for manipulating KeyStores.
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class KeyStoreUtil {
+
+    private static final String BCFKS = "BCFKS";
+    private static final String BKS = "BKS";
+    private static final String JCEKS = "JCEKS";
+    private static final String JKS = "JKS";
+    private static final String PKCS12 = "PKCS12";
+    private static final String UBER = "UBER";
+
+    private static final int VERSION_0 = 0;
+    private static final int VERSION_1 = 1;
+    private static final int VERSION_2 = 2;
+    private static final int JCEKS_MAGIC = 0xcececece;
+    private static final int JKS_MAGIC = 0xfeedfeed;
+    private static final int SEQUENCE = 0x30000000;
+
+    /**
+     * Tries to parse a keystore based on known recognizable patterns.
+     *
+     * This method can parse JKS, JCEKS, PKCS12, BKS, BCFKS and UBER key stores.
+     * At first the method looks for recognizable patterns of JKS, JCEKS, PKCS12 and BKS key store types and tries to parse them if found.
```

Comment:
Based on https://github.com/wildfly/wildfly-proposals/pull/124/commits/a313b5bd33b4aee706416550c1d1227ffbc69ec2#diff-02287f425f424b348e7faba51ae7acebR54 filename extension should be also tried.